README

	>>>> V E R S I O N    2  <<<<

INTRODUCTION:
Rh was written by Ken Stauffer to make the
job of finding files easier by allowing the
user to enter real C expressions. This notation is
much easier to master than the notation used by the
find(1) command, because most Unix users
already know C. In addition to being easier to use
than find(1), rh expressions can be used to select
the desired files.

CREDITS:
    Ken Stauffer
	- Original author.
    Guy Harris 
	- Corrected many portability problems.
    David MacKenzie 
	- Manual revisions. Added getopt and glob matching.
    Norm Hutchinson
	- Fixed ungetit().
    raf
	- Restored from oblivion after 32 years in the wilderness.
	- Updated to use ANSI/ISO C and POSIX.
	- Fixed buffer overflow security bugs and usability problems.

INSTALLING:
To install rh on your system:

    make
    make install

RUNNING:
There is a file called rhrc. This file contains some
examples of things that can go into a $HOME/.rhrc file.
If the file "rhrc" is moved to your home directory and renamed
to ".rhrc" then a command like:

	% rh -l -e writable

Will do a search of the current directory, executing the function
"writable()", which finds files that other people have write access to.

Once rh is made, you can do what you want with it. A good test to
see if it is working is to do:

	% rh -vle 1 /

This will find all files that make the constant expression '1' true.
So if you're root, all the files on the system will be found.

PORTABILITY:
Rh is written in ANSI/ISO C and should work on any POSIX system.

EXAMPLES:
Here are some example functions that can be defined in your ~/.rhrc file:

	dir() { return (mode & IFMT) == IFDIR; }
	months { return days * 30; }
	nlinks { return nlink; }
	writable() { return mode & 022; }
	mine { return uid == $$; }
	freds { return uid == $fred; }
	csrc() { return "*.c" || "*.h" || "[Mm]akefile"; }
	ago(d) { return NOW - d; }
	changed() { return mtime > ago(1 * hours); }
	K() { return 1024; }
	megs { return K << 10; }
	recursion(n) { return (n > 0) ? recursion(n - 1) + 1 : 0; }

GRAMMER:
The following is the grammer that describes the input language
recognized by rh:

	<program> ==> <function list> <expression> EOF
		| <function list> <expression> ;

	<function list> ==> <function>
		| <function list> <function>
		| /* empty */

	<function> ==> <function heading> { RETURN <expression> ; }

	<function heading> ==> IDENTIFIER
		| IDENTIFIER ( )
		| IDENTIFIER ( <idlist> )

	<idlist> ==> IDENTIFIER <idtail>
	<idtail> ==> , <idlist>
		| /* empty */
	
	<expression> ==> <expr0> ? <expression> : <expression>

	<expr0> ==> <expr1> || <expr1>

	<expr1> ==> <expr2> && <expr2>

	<expr2> ==> <expr3> | <expr3>

	<expr3> ==> <expr4> ^ <expr4>

	<expr4> ==> <expr5> & <expr5>

	<expr5> ==> <expr6> == <expr6>
		|   <expr6> != <expr6>

	<expr6> ==> <expr7> < <expr7>
		|   <expr7> > <expr7>
		|   <expr7> <= <expr7>
		|   <expr7> >= <expr7>

	<expr7> ==> <expr8> >> <expr8>
		|   <expr8> << <expr8>

	<expr8> ==> <expr9> + <expr9>
		|   <expr9> - <expr9>

	<expr9> ==> <expr10> * <expr10>
		|   <expr10> / <expr10>
		|   <expr10> % <expr10>

	<expr10> ==> ~ <expr10>
		|    ! <expr10>
		|    - <expr10>
		|    <factor>

	<factor> ==> ( <expression> )
		|    NUMBER
		|    <function call>
		|    IDENTIFIER
		|    [ <date spec> ]
		|    STRING

	<function call> ==> IDENTIFIER
		| IDENTIFIER ( <exprlist> )
		| IDENTIFIER ( )

	<exprlist> ==> <expression> <exprtail>
	<exprtail> ==> , <exprlist>
		| /* empty */

	<datespec> ==> NUMBER / NUMBER / NUMBER

--------------------------------------------------------------------
Ken Stauffer.
root@sixk


CHANGELOG
3.3 (20231013)

    - Makefile - Place env CFLAGS after explicit flags so they can override the explicit flags
    - configure - Fix --prefix= buglet (only worked if it was the first argument)
    - Makefile - Include env CPPFLAGS (just in case)
    - configure - Add --macports-prefix for macports with non-default prefix
    - configure - Add missing --etcdir and fix --destdir and --mandir
    - tests - Adapt to identify new adduser/addgroup help message
    - Add body pattern modifiers: .body .ibody .rebody .reibody
    - .reuse/dep5 - Make more accurate (strlcpy/strlcat not exactly BSD-3-Clause)
    - pcre2 - Stop using the PCRE2_NEVER_BACKSLASH_C option
    - configure - Remove 'when present' from pcre2 message for consistency
    - configure/tests - Replace `which` with `command -v` (more consistent)
    - tests - Make RAWHIDE_TEST_MULTIBYTE_USER_GROUP message shorter
    - tests - On FreeBSD chflags uappnd fails on zfs so test with sappnd
    - Add -j option (same as -L "%j\n")
    - configure - Add --enable-ubsan --enable-asan (undefined behaviour/address sanitizers)
    - doc - Show explicit -f- (as distinct from: -f fname) in --help output and SYNOPSIS
    - doc - Add rhs/rht/rhz etc. shell functions (rh/-r/-l/-v sorted (via jq) by path/mtime/size)
    - doc - Make rawhide.conf(5) manual entry fit in 80 columns again
    - doc - Update grammar to include recently added pattern modifiers
    - Fix -L %j access_control_list_verbose on FreeBSD/Solaris (wasn't verbose)
    - Revoke any setuid/setgid privileges at the start

3.2 (20230609)

    - Add support for attr (flags) on FreeBSD, OpenBSD, NetBSD, and macOS
    - Add ./configure --default (to avoid needing the details in Makefile)
    - Change --help pattern modifiers layout to improve readability
    - Remove explicit PCRE2 JIT compilation (had introduced uninitialized reads)
    - Add file type pattern modifiers: .what .iwhat .rewhat .reiwhat
    - Add MIME type pattern modifiers: .mime .imime .remime .reimime
    - Add -L %w (file type description) %W (MIME type)
    - Add to -L %j (JSON) output fields: filetype mimetype

3.1 (20221011)

    - Add built-in symbols for Linux ext2-style file attributes: attr proj gen
    - Add real /etc/rawhide.conf.d/attributes (for Linux with libe2p installed)
    - Add reference file fields: attr proj gen attribute project generation
    - Add -L %e %J %I conversions (Linux ext2-style file attributes project generation)
    - Add to -L %j (JSON) output, Linux ext2-style file attributes project generation
    - Add replacing "/".i "/".re "/".rei with implicit "/".ipath "/".repath "/".reipath
    - Add identifying pattern modifiers by unique prefixes (e.g., "/".p for "/".path)
    - Add "make default" target (extracted from make dist, to undo ./configure)
    - Fix -l column width buglet when user/group names contain multi-byte characters
    - Fix max path length buglet when crossing filesystem boundaries during traversal
    - Allow multiple -f options (but not multiple -f- because stdin can't be read twice)
    - Allow -f option argument to be a directory or to have a corresponding .d directory
    - Cache test-related getenv values that were being checked many times in normal use
    - Add a few more tests to maximize coverage within reason (97.04% on my Debian dev VM)
    - Add explicit PCRE2 JIT compilation (not noticeably faster - maybe default behaviour)

3.0 (20220330)

    - Update (after 32 years!) to compile and work again (ISO/IEC C, POSIX)
    - Remove internal getopt/glob_match, replacing with standard getopt/fnmatch
    - Remove MANIFEST (empty)
    - Remove META (only relevant for the original comp.source.unix release)
    - Add LICENSE COPYING CHANGELOG CODE_OF_CONDUCT.md
    - Change -h option so it terminates (was documented behaviour but still wrong)
    - Fix insecure multiple potential buffer overruns
    - Fix insecure -x shell command injection
    - Improve syntax error messages (more informative)
    - Increase static code/data/stack space by three orders of magnitude
    - Increase identifier length limit to 200
    - Add @groupname and @@ expression notation
    - Add support for underscore in function/parameter names
    - Add support for dot, dash, and underscore in user/group names
    - Add support for ~/.rhrc.d/*, /etc/rawhide.conf, and /etc/rawhide.conf.d/*
    - Add -N and -n options to suppress system-wide config and user-specific config
    - Add // and # single-line comment styles
    - Add built-in stat field symbols: blksize blocks
    - Change built-in symbols: hours/days/weeks to hour/day/week
    - Add built-in number symbols: second minute month year
    - Change to default "-e 1" expression to match everything (rather than stdin)
    - Add "-f -" convention to read from stdin (now that it's no longer the default)
    - Add error reporting during directory traversal (stderr and exit status)
    - Interpret two-digit years as being in the current century
    - Add built-in number symbols: IRWXU IRUSR IWUSR IXUSR
    - Add built-in number symbols: IRWXG IRGRP IWGRP IXGRP
    - Add built-in number symbols: IRWXO IROTH IWOTH IXOTH
    - Add /etc/rawhide.conf "standard" library (nice names for things, handy functions)
    - Remove example rhrc (subsumed by /etc/rawhide.conf)
    - Change built-in symbol NOW to now (not backwards compatible!) NOW { now }
    - Make "return" and ";" optional in functions (always redundant, more compact)
    - Make rh exit with failure status if any -x command exits with failure status
    - Add many ruthless tests
    - Add -V (version) option
    - Add --help and --version options
    - Add -0 option (nul-separated output) for use with xargs -0
    - Make "-e" optional for lean flexible cmdline (expr can appear among dir args)
    - Fix built-in prune over-pruning (wasn't reset until after siblings done)
    - Add built-in symbol: exit (terminate immediately - usage: 'expr ? exit : 0')

rhdata.c
/*
* rawhide - find files using pretty C expressions
* https://raf.org/rawhide
* https://github.com/raforg/rawhide
* https://codeberg.org/raforg/rawhide
*
* Copyright (C) 1990 Ken Stauffer, 2022-2023 raf <raf@raf.org>
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program; if not, see <https://www.gnu.org/licenses/>.
*
* 20231013 raf <raf@raf.org>
*/

#define _GNU_SOURCE /* For FNM_EXTMATCH and FNM_CASEFOLD in <fnmatch.h> */
#define _FILE_OFFSET_BITS 64 /* For 64-bit off_t on 32-bit systems (Not AIX) */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <fnmatch.h>
#include <time.h>
#include <sys/stat.h>

#define DATA
#include "rh.h"
#include "rhdata.h"
#include "rhcmds.h"
#include "rherr.h"

char *prog_name;                     /* Name of this program for messages */

symbol_t *symbols;                   /* Symbol table */
symbol_t *tokensym;                  /* Current token symbol */
llong tokenval;                      /* Current token value */
llong token;                         /* Current token code */

instr_t Program[MAX_PROGRAM_SIZE];   /* Program instructions */
llong PC;                            /* Program counter */
llong startPC;                       /* Initial program counter */

char Strbuf[MAX_DATA_SIZE];          /* Storage for string literals */
llong strfree = 0;                   /* Index into Strbuf for the next string literal */

reffile_t RefFile[MAX_REFFILE_SIZE]; /* Storage for reference files */
llong reffree = 0;                   /* Index into RefFile for the next reference file */

llong Stack[MAX_STACK_SIZE + 3];     /* Stack */
llong SP;                            /* Stack pointer */
llong FP;                            /* Frame pointer */

runtime_t attr;                      /* Configuration and runtime state */

char *expstr;   /* The -e option argument (search criteria expression) */
char *expfname; /* The -f option argument (filename) or current config file */
FILE *expfile;  /* FILE handle corresponding to the -f option argument or current config file */

#ifndef S_IFDOOR
#define S_IFDOOR 0150000
#endif

/* Initial built-in symbol table */

static symbol_t init_syms[] =
{
	/* stat structure fields */

	{ "dev",     FIELD, 0, c_dev,     NULL },
	{ "major",   FIELD, 0, c_major,   NULL },
	{ "minor",   FIELD, 0, c_minor,   NULL },
	{ "ino",     FIELD, 0, c_ino,     NULL },
	{ "mode",    FIELD, 0, c_mode,    NULL },
	{ "nlink",   FIELD, 0, c_nlink,   NULL },
	{ "uid",     FIELD, 0, c_uid,     NULL },
	{ "gid",     FIELD, 0, c_gid,     NULL },
	{ "rdev",    FIELD, 0, c_rdev,    NULL },
	{ "rmajor",  FIELD, 0, c_rmajor,  NULL },
	{ "rminor",  FIELD, 0, c_rminor,  NULL },
	{ "size",    FIELD, 0, c_size,    NULL },
	{ "blksize", FIELD, 0, c_blksize, NULL },
	{ "blocks",  FIELD, 0, c_blocks,  NULL },
	{ "atime",   FIELD, 0, c_atime,   NULL },
	{ "mtime",   FIELD, 0, c_mtime,   NULL },
	{ "ctime",   FIELD, 0, c_ctime,   NULL },

	/* Linux ext2-style file attributes and BSD file flags */

#if HAVE_ATTR || HAVE_FLAGS
	{ "attr",    FIELD, 0, c_attr,   NULL },
#endif
#if HAVE_ATTR
	{ "proj",    FIELD, 0, c_proj,   NULL },
	{ "gen",     FIELD, 0, c_gen,    NULL },
#endif

	/* Miscellaneous functions and control flow */

	{ "nouser",     FIELD, 0, c_nouser,     NULL },
	{ "nogroup",    FIELD, 0, c_nogroup,    NULL },
	{ "readable",   FIELD, 0, c_readable,   NULL },
	{ "writable",   FIELD, 0, c_writable,   NULL },
	{ "executable", FIELD, 0, c_executable, NULL },
	{ "strlen",     FIELD, 0, c_strlen,     NULL },
	{ "depth",      FIELD, 0, c_depth,      NULL },
	{ "prune",      FIELD, 0, c_prune,      NULL },
	{ "trim",       FIELD, 0, c_trim,       NULL },
	{ "exit",       FIELD, 0, c_exit,       NULL },

	/* Time */

	{ "now",    NUMBER, 0,                            c_number, NULL }, /* Set by rawhide_init() */
	{ "today",  NUMBER, 0,                            c_number, NULL }, /* Set by rawhide_init() */
	{ "second", NUMBER, 1,                            c_number, NULL },
	{ "minute", NUMBER, 60,                           c_number, NULL },
	{ "hour",   NUMBER, 60 * 60,                      c_number, NULL },
	{ "day",    NUMBER, 60 * 60 * 24,                 c_number, NULL },
	{ "week",   NUMBER, 60 * 60 * 24 * 7,             c_number, NULL },
	{ "month",  NUMBER, 60 * 60 * 24 * 365.2425 / 12, c_number, NULL },
	{ "year",   NUMBER, 60 * 60 * 24 * 365.2425,      c_number, NULL },

	/* Constants from <sys/stat.h> */

	{ "IFREG",  NUMBER, S_IFREG,  c_number, NULL },
	{ "IFDIR",  NUMBER, S_IFDIR,  c_number, NULL },
	{ "IFLNK",  NUMBER, S_IFLNK,  c_number, NULL },
	{ "IFCHR",  NUMBER, S_IFCHR,  c_number, NULL },
	{ "IFBLK",  NUMBER, S_IFBLK,  c_number, NULL },
	{ "IFSOCK", NUMBER, S_IFSOCK, c_number, NULL },
	{ "IFIFO",  NUMBER, S_IFIFO,  c_number, NULL },
	{ "IFDOOR", NUMBER, S_IFDOOR, c_number, NULL },
	{ "IFMT",   NUMBER, S_IFMT,   c_number, NULL },
	{ "ISUID",  NUMBER, S_ISUID,  c_number, NULL },
	{ "ISGID",  NUMBER, S_ISGID,  c_number, NULL },
	{ "ISVTX",  NUMBER, S_ISVTX,  c_number, NULL },
	{ "IRWXU",  NUMBER, S_IRWXU,  c_number, NULL },
	{ "IRUSR",  NUMBER, S_IRUSR,  c_number, NULL },
	{ "IWUSR",  NUMBER, S_IWUSR,  c_number, NULL },
	{ "IXUSR",  NUMBER, S_IXUSR,  c_number, NULL },
	{ "IRWXG",  NUMBER, S_IRWXG,  c_number, NULL },
	{ "IRGRP",  NUMBER, S_IRGRP,  c_number, NULL },
	{ "IWGRP",  NUMBER, S_IWGRP,  c_number, NULL },
	{ "IXGRP",  NUMBER, S_IXGRP,  c_number, NULL },
	{ "IRWXO",  NUMBER, S_IRWXO,  c_number, NULL },
	{ "IROTH",  NUMBER, S_IROTH,  c_number, NULL },
	{ "IWOTH",  NUMBER, S_IWOTH,  c_number, NULL },
	{ "IXOTH",  NUMBER, S_IXOTH,  c_number, NULL },

	/* Symlink target stat structure fields */

	{ "texists",   FIELD, 0, t_exists,  NULL },
	{ "tdev",      FIELD, 0, t_dev,     NULL },
	{ "tmajor",    FIELD, 0, t_major,   NULL },
	{ "tminor",    FIELD, 0, t_minor,   NULL },
	{ "tino",      FIELD, 0, t_ino,     NULL },
	{ "tmode",     FIELD, 0, t_mode,    NULL },
	{ "tnlink",    FIELD, 0, t_nlink,   NULL },
	{ "tuid",      FIELD, 0, t_uid,     NULL },
	{ "tgid",      FIELD, 0, t_gid,     NULL },
	{ "trdev",     FIELD, 0, t_rdev,    NULL },
	{ "trmajor",   FIELD, 0, t_rmajor,  NULL },
	{ "trminor",   FIELD, 0, t_rminor,  NULL },
	{ "tsize",     FIELD, 0, t_size,    NULL },
	{ "tblksize",  FIELD, 0, t_blksize, NULL },
	{ "tblocks",   FIELD, 0, t_blocks,  NULL },
	{ "tatime",    FIELD, 0, t_atime,   NULL },
	{ "tmtime",    FIELD, 0, t_mtime,   NULL },
	{ "tctime",    FIELD, 0, t_ctime,   NULL },
	{ "tstrlen",   FIELD, 0, t_strlen,  NULL },

	/* Reference file stat structure fields */

	{ ".exists",   REFFILE, 0, r_exists,  NULL },
	{ ".dev",      REFFILE, 0, r_dev,     NULL },
	{ ".major",    REFFILE, 0, r_major,   NULL },
	{ ".minor",    REFFILE, 0, r_minor,   NULL },
	{ ".ino",      REFFILE, 0, r_ino,     NULL },
	{ ".mode",     REFFILE, 0, r_mode,    NULL },
	{ ".type",     REFFILE, 0, r_type,    NULL },
	{ ".perm",     REFFILE, 0, r_perm,    NULL },
	{ ".nlink",    REFFILE, 0, r_nlink,   NULL },
	{ ".uid",      REFFILE, 0, r_uid,     NULL },
	{ ".gid",      REFFILE, 0, r_gid,     NULL },
	{ ".rdev",     REFFILE, 0, r_rdev,    NULL },
	{ ".rmajor",   REFFILE, 0, r_rmajor,  NULL },
	{ ".rminor",   REFFILE, 0, r_rminor,  NULL },
	{ ".size",     REFFILE, 0, r_size,    NULL },
	{ ".blksize",  REFFILE, 0, r_blksize, NULL },
	{ ".blocks",   REFFILE, 0, r_blocks,  NULL },
	{ ".atime",    REFFILE, 0, r_atime,   NULL },
	{ ".mtime",    REFFILE, 0, r_mtime,   NULL },
	{ ".ctime",    REFFILE, 0, r_ctime,   NULL },
#if HAVE_ATTR || HAVE_FLAGS
	{ ".attr",     REFFILE, 0, r_attr,    NULL },
#endif
#if HAVE_ATTR
	{ ".proj",     REFFILE, 0, r_proj,    NULL },
	{ ".gen",      REFFILE, 0, r_gen,     NULL },
#endif
	{ ".strlen",   REFFILE, 0, r_strlen,  NULL },

	{ ".inode",    REFFILE, 0, r_ino,     NULL },
	{ ".nlinks",   REFFILE, 0, r_nlink,   NULL },
	{ ".user",     REFFILE, 0, r_uid,     NULL },
	{ ".group",    REFFILE, 0, r_gid,     NULL },
	{ ".sz",       REFFILE, 0, r_size,    NULL },
	{ ".accessed", REFFILE, 0, r_atime,   NULL },
	{ ".modified", REFFILE, 0, r_mtime,   NULL },
	{ ".changed",  REFFILE, 0, r_ctime,   NULL },
#if HAVE_ATTR || HAVE_FLAGS
	{ ".attribute",  REFFILE, 0, r_attr,  NULL },
#endif
#if HAVE_ATTR
	{ ".project",    REFFILE, 0, r_proj,  NULL },
	{ ".generation", REFFILE, 0, r_gen,   NULL },
#endif
	{ ".len",      REFFILE, 0, r_strlen,  NULL },

	/* Pattern modifiers */

#ifdef FNM_CASEFOLD
	{ ".i",        PATMOD, 0, c_i,       NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".re",       PATMOD, 0, c_re,      NULL },
	{ ".rei",      PATMOD, 0, c_rei,     NULL },
#endif

	{ ".path",     PATMOD, 0, c_path,    NULL },
#ifdef FNM_CASEFOLD
	{ ".ipath",    PATMOD, 0, c_ipath,   NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".repath",   PATMOD, 0, c_repath,  NULL },
	{ ".reipath",  PATMOD, 0, c_reipath, NULL },
#endif

	{ ".link",     PATMOD, 0, c_link,    NULL },
#ifdef FNM_CASEFOLD
	{ ".ilink",    PATMOD, 0, c_ilink,   NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".relink",   PATMOD, 0, c_relink,  NULL },
	{ ".reilink",  PATMOD, 0, c_reilink, NULL },
#endif

	{ ".body",     PATMOD, 0, c_body,    NULL },
#ifdef FNM_CASEFOLD
	{ ".ibody",    PATMOD, 0, c_ibody,   NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".rebody",   PATMOD, 0, c_rebody,  NULL },
	{ ".reibody",  PATMOD, 0, c_reibody, NULL },
#endif

#ifdef HAVE_MAGIC
	{ ".what",     PATMOD, 0, c_what,     NULL },
#ifdef FNM_CASEFOLD
	{ ".iwhat",    PATMOD, 0, c_iwhat,    NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".rewhat",   PATMOD, 0, c_rewhat,   NULL },
	{ ".reiwhat",  PATMOD, 0, c_reiwhat,  NULL },
#endif
#endif

#ifdef HAVE_MAGIC
	{ ".mime",     PATMOD, 0, c_mime,     NULL },
#ifdef FNM_CASEFOLD
	{ ".imime",    PATMOD, 0, c_imime,    NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".remime",   PATMOD, 0, c_remime,   NULL },
	{ ".reimime",  PATMOD, 0, c_reimime,  NULL },
#endif
#endif

#ifdef HAVE_ACL
	{ ".acl",      PATMOD, 0, c_acl,     NULL },
#ifdef FNM_CASEFOLD
	{ ".iacl",     PATMOD, 0, c_iacl,    NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".reacl",    PATMOD, 0, c_reacl,   NULL },
	{ ".reiacl",   PATMOD, 0, c_reiacl,  NULL },
#endif
#endif

#ifdef HAVE_EA
	{ ".ea",       PATMOD, 0, c_ea,      NULL },
#ifdef FNM_CASEFOLD
	{ ".iea",      PATMOD, 0, c_iea,     NULL },
#endif
#ifdef HAVE_PCRE2
	{ ".reea",     PATMOD, 0, c_reea,    NULL },
	{ ".reiea",    PATMOD, 0, c_reiea,   NULL },
#endif
#endif
	{ ".sh",       PATMOD, 0, c_sh,      NULL },
};

/*

void rawhide_init(void);

Prepare the symbol table with built-in symbols.
Set the values for "now" and "today".
Must be called before the -h option is processed.

*/

void rawhide_init(void)
{
	symbol_t *sym;
	struct tm *tm;
	time_t t;
	int i;

	symbols = &init_syms[0];

	for (i = 0; i < sizeof(init_syms) / sizeof(symbol_t) - 1; i++)
		init_syms[i].next = &init_syms[i + 1];

	/* Initialize "now" to the time right now */

	sym = locate_symbol("now");
	sym->value = (llong)time(&t);

	/* Initialize "today" to the time last midnight (local time) */

	tm = localtime(&t);
	tm->tm_hour = tm->tm_min = tm->tm_sec = 0;
	tm->tm_isdst = -1;
	t = mktime(tm);

	sym = locate_symbol("today");
	sym->value = (llong)t;

	/* Initialize Program */

	PC = 0;
	startPC = -1;
}

/*

void rawhide_finish(void);

Free non-built-in symbols.
Must be called after parsing.
Can be called before searching.

*/

void rawhide_finish(void)
{
	symbol_t *s;

	while (symbols->type == PARAM || symbols->type == FUNCTION || symbols->type == IDENTIFIER)
	{
		s = symbols;
		symbols = symbols->next;
		free(s->name);
		free(s);
	}
}

/*

symbol_t *insert_symbol(char *name, int toktype, llong val);

Insert a new symbol into the symbol table.
The name parameter is its name.
The toktype parameter is its type.
The val parameter is its value, or zero.

Return a pointer to the symbol table entry.
The symbol is inserted at the head of the linked list.
This behaviour is relied upon elsewhere.

*/

symbol_t *insert_symbol(char *name, int toktype, llong val)
{
	symbol_t *sym;

	if (!(sym = malloc(sizeof(symbol_t))))
		return NULL;

	if (!(sym->name = strdup(name)))
	{
		free(sym);

		return NULL;
	}

	sym->type = toktype;
	sym->value = val;
	sym->next = symbols;
	symbols = sym;

	return sym;
}

/*

symbol_t *locate_symbol(char *name);

Search for a symbol in the symbol table.
The name parameter is the name of the symbol to search for.

*/

symbol_t *locate_symbol(char *name)
{
	symbol_t *s;

	for (s = symbols; s; s = s->next)
		if (!strcmp(name, s->name))
			return s;

	return NULL;
}

/*

symbol_t *locate_patmod_prefix(char *name);

Search for a pattern modifier symbol in the symbol table.
The name parameter is hopefully a unique prefix of a symbol,
but it is not a complete symbol.

*/

symbol_t *locate_patmod_prefix(char *name)
{
	symbol_t *s, *match = NULL;

	for (s = symbols; s; s = s->next)
	{
		if (s->type != PATMOD)
			continue;

		if (!strncmp(name, s->name, strlen(name)))
		{
			if (match)
				return NULL; /* Not a unique prefix */

			match = s;
		}
	}

	return match;
}

/*

int rawhide_instruction(void (*func)(llong), llong value);

Append a rawhide instruction to the Program array.
The func parameter is the instruction's implementation.
The value parameter is the argument to that function.

*/

int rawhide_instruction(void (*func)(llong), llong value)
{
	if (PC >= MAX_PROGRAM_SIZE)
		return -1;

	Program[PC].func = func;
	Program[PC++].value = value;

	return 0;
}

/*

llong rawhide_execute(void);

Execute the program stored in Program.
Each element of Program contains a pointer to a function.
The program is NULL-terminated.
Returns the value of the expression.

*/

llong rawhide_execute(void)
{
	for (SP = 0, PC = startPC; Program[PC].func; PC++)
	{
		(*Program[PC].func)(Program[PC].value);

		if (SP >= MAX_STACK_SIZE)
			fatal("stack overflow");
	}

	return Stack[0];
}

/* vi:set ts=4 sw=4: */
